cirq/google/serializable_device.py

```diff
@@ -0,0 +1,136 @@
+        print(self.durations)
+
+    def _qubits_from_ids(self, id_list):
+        """Translates a list of ids in proto format e.g. '4_3'
+        into cirq.GridQubit objects"""
+        # TODO(dstrain): Add support for non-grid qubits
+        return [devices.GridQubit.from_proto_id(id) for id in id_list]
+
+    def _create_target_set(self, ts: device_pb2.TargetSet) -> Set[Tuple]:
+        """Transform a TargetSet proto into a set of qubit tuples"""
+        # TODO(dstrain): add support for measurement qubits
+        target_set = set()
+        for target in ts.targets:
+            qid_list = self._qubits_from_ids(target.ids)
+            target_set.add(tuple(qid_list))
+            if ts.target_ordering == device_pb2.TargetSet.SYMMETRIC:
+                qid_list.reverse()
+                target_set.add(tuple(qid_list))
+        return target_set
+
+    def _find_operation_type(self, op_key: 'cirq.Operation',
+                             type_dict: Dict[Any, Any]):
+        """Finds the type (or a compatible type) of an operation from within
+        a dictionary with keys of Gate type.
+
+        Returns:
+             the value corresponding to that key or None if no type matches
+        """
+        gate_op = cast(ops.GateOperation, op_key)
+        gate_key = gate_op.gate
+        if gate_key in type_dict:
+            return type_dict[gate_key]
+        for type_key in type_dict:
+            if isinstance(gate_key, type_key):
+                return type_dict[type_key]
+        return None
+
+    def duration_of(self, operation: 'cirq.Operation') -> Duration:
+        duration = self._find_operation_type(operation, self.durations)
+        if duration is not None:
+            return duration
+        raise ValueError(
+            f'Operation {operation} does not have a duration listed')
+
+    def validate_operation(self, operation: 'cirq.Operation') -> None:
+        for q in operation.qubits:
+            if q not in self.qubits:
+                raise ValueError('Qubit not on device: {!r}'.format(q))
+
+        if (len(operation.qubits) > 1):
+            # TODO(dstrain): verify number of qubits and args
+
+            qubit_tuple = tuple(operation.qubits)
+            gate_op = cast(ops.GateOperation, operation)
+            for t in self.target_sets:
+                if isinstance(gate_op.gate, t):
+                    for ts in self.target_sets[t]:
+                        if qubit_tuple in self.allowed_targets[ts]:
+                            # Valid
+                            return
+            # Target is not within any of the target sets specified by the gate.
+            raise ValueError(
+                f'Operation does not use valid qubit target: {operation}.')
+
+    def validate_scheduled_operation(
+            self, schedule: 'cirq.Schedule',
+            scheduled_operation: 'cirq.ScheduledOperation') -> None:
+        pass
+
+    def validate_schedule(self, schedule: 'cirq.Schedule') -> None:
+        pass
```

Comment:
`raise NotImplementedError`?

Reply:
I think that would render the class useless, since every schedule would be considered invalid then.  I will add this in a follow-up PR unless I can finish while people are still reviewing.

Reply:
got it